src/main/java/de/zalando/aruha/nakadi/service/subscription/state/ClosingState.java

```diff
@@ -0,0 +1,146 @@
+package de.zalando.aruha.nakadi.service.subscription.state;
+
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZKSubscription;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ClosingState extends State {
+    private final Map<Partition.PartitionKey, Long> uncommitedOffsets;
+    private final Map<Partition.PartitionKey, ZKSubscription> listeners = new HashMap<>();
+    private final long lastCommitMillis;
+    private ZKSubscription topologyListener;
+    private static final Logger LOG = LoggerFactory.getLogger(ClosingState.class);
+
+    ClosingState(final Map<Partition.PartitionKey, Long> uncommitedOffsets, final long lastCommitMillis) {
+        this.uncommitedOffsets = uncommitedOffsets;
+        this.lastCommitMillis = lastCommitMillis;
+    }
+
+    @Override
+    public void onExit() {
+        try {
+            freePartitions(new HashSet<>(listeners.keySet()));
+        } finally {
+            if (null != topologyListener) {
+                try {
+                    topologyListener.cancel();
+                } finally {
+                    topologyListener = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void onEnter() {
+        final long timeToWaitMillis = getParameters().commitTimeoutMillis - (System.currentTimeMillis() - lastCommitMillis);
+        if (timeToWaitMillis > 0) {
+            scheduleTask(() -> {
+                if (isCurrent()) {
+                    switchState(new CleanupState());
+                }
+            }, timeToWaitMillis, TimeUnit.MILLISECONDS);
+            topologyListener = getZk().subscribeForTopologyChanges(() -> addTask(this::onTopologyChanged));
+            reactOnTopologyChange();
+        } else {
+            switchState(new CleanupState());
+        }
+    }
+
+    private void onTopologyChanged() {
+        if (!isCurrent()) {
+            return;
+        }
+        topologyListener.refresh();
+        reactOnTopologyChange();
+    }
+
+    private void reactOnTopologyChange() {
+        final Map<Partition.PartitionKey, Partition> partitions = new HashMap<>();
+        getZk().runLocked(() -> Stream.of(getZk().listPartitions())
+                .filter(p -> getSessionId().equals(p.getSession()))
+                .forEach(p -> partitions.put(p.getKey(), p)));
+        final Set<Partition.PartitionKey> freeRightNow = new HashSet<>();
+        final Set<Partition.PartitionKey> addListeners = new HashSet<>();
+        for (final Partition p : partitions.values()) {
```

Comment:
I think one thing that could improve readability would be to split it in many scenarios. Even if that means iterating multiple times over the collection. Something like:

1. freeCommitedPartitions
2. freeNotAssignedPartitions
3. ...

and so on. Then write some comments for each method explaining the necessity for it. For someone reading this code for the first time, it handles so many scenarios that it's almost impossible for me to tell why or if is there any missing scenario to be handled.